get-runs interaction

xanderflood opened the pull request:
Extending the `pkg/postgtes.Repository` interface to include a `GetRecentlyActiveRuns` method which joins up all the tables and returns a nested structure for looking up train state

## Tested:
Tested using [this main](https://gist.github.com/xanderflood/fd4c0f253f28c6b2aa702690dd24bf09) with a connection to the scrapedumper postgres DB and got [this output](https://gist.github.com/xanderflood/49c6945aeb5083614e0154239f83a59b).

xanderflood commented on pkg/postgres/run.go:
```diff
@@ -0,0 +1,48 @@
+	r.TrainID = parts[2]
+}
+
+func (r Run) Finished() bool {
```
This can be used by the websocket to interpret the changes and decide when a run can be removed from the state. Once an arrival time is set for the terminus of a line, no further updates will be made to the record, so it won't come up again in future polls of `GetRecentlyActiveRuns`

xanderflood commented on pkg/postgres/repository.go:
```diff
@@ -327,6 +329,87 @@ WHERE most_recent_event_moment < $1`,
 	return
 }
 
+func (a *RepositoryAgent) GetRecentlyActiveRuns(touchThreshold EasternTime) (runs map[martaapi.Line]map[martaapi.Direction][]Run, err error) {
```
Collects all the data about any runs that have been updated since `touchThreshold`

xanderflood replied:
If this query is slow, we can select fewer columns and add some indices. Right now that doesn't appear necessary (it's taking about 30ms against the live data 🤷‍♀)

bipol replied:
nother nit: would it be worth adding your first comment as a doc comment for the function?

xanderflood commented on pkg/postgres/time.go:
```diff
@@ -46,5 +46,6 @@ func (ae *EasternTime) Scan(value interface{}) error {
 
 //Value implements the db/sql.Valuer interface
 func (ae EasternTime) Value() (driver.Value, error) {
+	ae = EasternTime(time.Time(ae).In(EasternTimeZone))
```
This wasn't a problem because all the `time.Time`'s we _had_ been passing into SQL queries were already in Eastern. Adding this line doesn't change anything but helps make sure that `GetRecentlyActiveRuns` works properly no matter what `time.Location` the `touchThreshold` argument is in.

bipol commented on pkg/martaapi/taxonomy.go:
```diff
@@ -209,3 +204,23 @@ var LineDirections = map[Line][]Direction{
+//Termini allow for lookups of all the terminuseses of the different lines
+var Termini = map[Line]map[Direction]Station{
+	Green: map[Direction]Station{
+		East: EdgewoodCandlerParkStation,
```
reel dum nitpick:  For the `Station` type, we could remove the word `Station` each `const` to reduce stuttering

bipol commented on pkg/postgres/repository.go:
```diff
@@ -327,6 +329,87 @@ WHERE most_recent_event_moment < $1`,
+		run.setLineDirectionAndTrainID()
+
+		if seenRun, ok := runsByIdentifier[run.Identifier]; ok {
+			run = seenRun
```
Since we reset the run here if we've seen it before, do we need to run line 377 run.setLineDirectionAndTrainID() before this? Not really an optimization or anything, just making sure I'm following the code right

bipol commented on pkg/postgres/repository.go:
```diff
@@ -327,6 +329,87 @@ WHERE most_recent_event_moment < $1`,
+		if seenRun, ok := runsByIdentifier[run.Identifier]; ok {
+			run = seenRun
+		} else {
+			run.Arrivals = map[martaapi.Station]Arrival{}
```
could this be handled in `setLineAndTrainID` as well?  maybe we could rename that function as well to an "initialize" or something else.  It may make more sense to stay here for clarity!

xanderflood commented:
@bipol your comments made me notice some unnecessary crap, so I made a couple little changes:

First of all, parsing the line and direction out of the identifier was dumb: we already have the corrected_line and corrected_direction fields, and those have better values. We also parsed the train_id out of the identifier, which we don't currently use, so I removed those three entirely to simplify things. If we ever want the train_id, we can look at adding that back.

I also dropped all that "group by line and direction" crap because this is DB-layer code and why make assumptions about how the consumer will want it grouped and arranged 🤷‍♀

jbonzo commented on pkg/martaapi/taxonomy.go:
```diff
@@ -209,3 +204,23 @@ var LineDirections = map[Line][]Direction{
+
+//Termini allow for lookups of all the terminuseses of the different lines
+var Termini = map[Line]map[Direction]Station{
+	Green: map[Direction]Station{
```
not saying this should be changed now, but this kind of universal information might best be stored somewhere more central since I could envision other repo's using this taxonomy. Maybe the gomarta client? or maybe a different repo. Thoughts?

xanderflood replied:
Yes Chris brought this up a while ago and I 100% agree, if you can make it to the meeting tomorrow let's chat about options for that

jbonzo commented on pkg/postgres/repository.go:
```diff
@@ -327,6 +329,76 @@ WHERE most_recent_event_moment < $1`,
+			&estimatedArrivalTime,
+		)
+		if err != nil {
+			err = errors.Wrapf(err, "failed to scan run")
```
ditto err wrap comment

jbonzo commented:
also im not familiar with your CI yet but its failing

xanderflood commented on pkg/postgres/repository.go:
```diff
@@ -327,6 +329,76 @@ WHERE most_recent_event_moment < $1`,
+		touchThreshold,
+	)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to get active runs")
```
👍 that reminds me I need to bump our go version in the .travis.yml as well - this is still building with 1.11
